### ISSUE TITLE
Training crashes with Background Improvement enabled – mean_step_far_mask is not CUDA device memory

### ISSUE THREAD

**Issue by Yutaka1111** [MrNeRF/LichtFeld-Studio#2069]
### Description

Training consistently fails when "Background Improvement" is enabled with the MRNF strategy.

The same datasets train successfully when Background Improvement is disabled.

I reproduced the issue with both:
- Dense initial point cloud
- Sparse COLMAP initial point cloud

So the issue does not appear to be specific to Dense initialization.

### Environment

LichtFeld Studio Nightly:
v0.5.3-519-g7555a559
commit: 7555a559

OS:
Windows

GPU:
NVIDIA RTX A6000
48 GB VRAM

CUDA driver reported by LichtFeld:
13.0

Training strategy:
MRNF

### Reproduction

1. Load a COLMAP dataset
2. Select MRNF
3. Enable "Background Improvement"
4. Start training
5. Training fails during optimization/refinement

The issue occurs with both Sparse and Dense initial point clouds.

If "Background Improvement" is disabled, the same Dense dataset completes training successfully.

I also tested reducing the initial point cloud / Gaussian count, but the problem still occurs.

### Error

The log consistently reports:

LFS FAILURE REPORT
Family: tensor contract violation
Contract: LFS boundary contract
Failed expression:
attributes.type == cudaMemoryTypeDevice

Context:
CUDA pointer 'mean_step_far_mask' has memory type 0 instead of device type 2

Detection site:
src/training/rasterization/fastgs/optimizer/src/adam_api.cu:173

The failure occurs inside:

adam_step_joint_contiguous_batched(...)

### Additional observations

Before the crash the log shows MRNF far-field processing is active, for example:

MRNF: deep-far census unconditional at threshold 0 (... far features active)
MRNF: far starvation 1.00 (far-field scene)

In some runs the crash occurs shortly after:

Morton reordered ... Gaussians

Examples from multiple runs:
- iteration 10234
- iteration 15351
- iteration 17850
- iteration 334
- around iteration 500 in another Sparse test

This makes the failure appear reproducible rather than a random CUDA OOM.

### Expected behavior

Training should complete normally when Background Improvement is enabled.

### Actual behavior

Training terminates with:

Training step failed:
LFS boundary contract failed:
attributes.type == cudaMemoryTypeDevice —
CUDA pointer 'mean_step_far_mask' has memory type 0 instead of device type 2

I can provide the full lichtfeld.log if needed.

**Comment by Yutaka1111** [comment]
Thanks for the fix and the detailed explanation.

Unfortunately, I tested today's latest Windows nightly build as well, and the same problem still occurs on my system.

Environment:
- Windows
- NVIDIA RTX A6000 48 GB
- MRNF
- Background Improvement enabled

Previously tested builds that reproduce the issue:
- 2026-09-07 — 8a6aed66
- 2026-09-05 — 7555a559
- 2026-09-04 — 94957666
- 2026-08-31 — 8e1a6ac2

I have now also tested the latest nightly available today (2026-09-09), and training still fails with Background Improvement enabled.

The same dataset completes training when Background Improvement is disabled.

I have also reproduced the issue with both:
- Dense initial point cloud
- Sparse COLMAP initial point cloud

So this does not appear to be specific to Dense initialization or the initial Gaussian count.

Since the Linux headless test completes successfully, this may be a Windows-specific issue or a remaining Windows CUDA/Vulkan interoperability path.

I will attach the latest lichtfeld.log from the 2026-09-09 nightly run.

**Comment by MrNeRF** [comment]
Thanks for retesting. The 2026-09-09 nightly (12802b8fa) was built a few hours before #2070 merged, so it does not contain that fix yet. Please try the 2026-09-10 nightly or later.

We also opened #2074: the optimizer now keeps the mask tensor alive itself and checks that it is device memory when it is bound, and MRNF rebuilds the mask when Background Improvement is switched on after start. If the failure still happens with a nightly that contains both, please attach the lichtfeld.log again.

**Comment by jacobvanbeets** [comment]
tested with BG improvement on. Commit: b6e0ee1e5
Worked for me.

<img width="1919" height="1030" alt="Image" src="https://github.com/user-attachments/assets/4239a3e2-801e-4c4c-b44b-1886015f96c4" />
